corehq/apps/locations/util.py

```diff
@@ -138,139 +108,143 @@ def lookup_by_property(domain, prop_name, val, scope, root=None):
-    except ResourceNotFound:
-        # property is not indexed
-        uniqueness_set = []
-        if scope == 'global':
-            uniqueness_set = [l for l in all_locations(loc.domain) if l._id != loc._id]
-        elif scope == 'siblings':
-            uniqueness_set = loc.siblings()
-
-        return set(l._id for l in uniqueness_set if val == normalize(getattr(l, prop_name, None)))
-
-
-def get_custom_property_names(domain, loc_type, common_types):
-    return [prop.name for prop in location_custom_properties(domain, loc_type) if prop.name not in common_types]
-
-
-def get_default_column_data(domain, location_types):
-    data = {
-        'headers': {},
-        'values': {}
-    }
-
-    if Domain.get_by_name(domain).commtrack_settings.individual_consumption_defaults:
-        products = Product.by_domain(domain)
-
-        supply_point_map = SupplyPointCase.get_location_map_by_domain(domain)
-
-        consumption_dict = build_consumption_dict(domain)
-
-        if not consumption_dict:
-            return data
-
-        for loc_type in location_types:
-            loc = get_loc_config(domain)[loc_type]
-            if not loc.administrative:
-                data['headers'][loc_type] = [
-                    'default_' +
-                    p.code for p in products
-                ]
-
-                locations = Location.filter_by_type(domain, loc_type)
-                for loc in locations:
-                    if loc._id in supply_point_map:
-                        sp_id = supply_point_map[loc._id]
-                    else:
-                        # this only happens if the supply point case did
-                        # not already exist
-                        sp_id = SupplyPointCase.get_or_create_by_location(loc)._id
-
-                    data['values'][loc._id] = [
-                        get_loaded_default_monthly_consumption(
-                            consumption_dict,
-                            domain,
-                            p._id,
-                            loc_type,
-                            sp_id
-                        ) or '' for p in products
-                    ]
-            else:
-                data['headers'][loc_type] = []
-    return data
 
+class LocationExporter(object):
+    def __init__(self, domain, include_consumption=False):
+        self.domain = domain
+        self.domain_obj = Domain.get_by_name(domain)
+        self.include_consumption_flag = include_consumption
+        self.data_model = get_location_data_model(domain)
+
+    @property
+    @memoized
+    def consumption_dict(self):
+        return build_consumption_dict(self.domain)
+
+    @property
+    @memoized
+    def include_consumption(self):
+        if bool(
+            self.include_consumption_flag and
+            self.domain_obj.commtrack_settings.individual_consumption_defaults
+        ):
+            # we'll be needing these, so init 'em:
```

Comment:
fyi this caused an issue: https://github.com/dimagi/commcare-hq/pull/6835

generally not a huge fan of this pattern since it makes other functions/callers need to be aware of whether the thing has been properly initialized or not. 

If you just make memoized properties for each of the things you might need then I think we can do all the calculations just in time and make the code more robust.